drivers/misc/ipts/ipts-companion.c

```diff
@@ -0,0 +1,203 @@
+#include <linux/firmware.h>
+#include <linux/ipts.h>
+#include <linux/ipts-binary.h>
+#include <linux/ipts-companion.h>
+#include <linux/mutex.h>
+
+#include "ipts.h"
+#include "ipts-companion.h"
+#include "ipts-params.h"
+
+#define IPTS_FW_PATH_FMT "intel/ipts/%s"
+#define IPTS_FW_CONFIG_FILE "ipts_fw_config.bin"
+
+ipts_companion_t *ipts_companion;
+DEFINE_MUTEX(ipts_companion_lock);
+
+bool ipts_companion_available(void)
+{
+	bool ret;
+	mutex_lock(&ipts_companion_lock);
+
+	ret = ipts_companion != NULL;
+
+	mutex_unlock(&ipts_companion_lock);
+	return ret;
+}
+
+/*
+ * General purpose API for adding or removing a companion driver
+ * A companion driver is a driver that implements hardware specific
+ * behaviour into IPTS, so it doesn't have to be hardcoded into the
+ * main driver. All requests to the companion driver should be wrapped,
+ * with a fallback in case a companion driver cannot be found.
+ */
+
+int ipts_add_companion(ipts_companion_t *companion)
+{
+	int ret = 0;
+	mutex_lock(&ipts_companion_lock);
+
+	if (ipts_companion != NULL) {
+		ret = -EBUSY;
+		goto add_companion_return;
+	}
+
+	ipts_companion = companion;
+
+add_companion_return:
+
+	mutex_unlock(&ipts_companion_lock);
+	return ret;
+}
+EXPORT_SYMBOL_GPL(ipts_add_companion);
+
+int ipts_remove_companion(ipts_companion_t *companion)
```

Comment:
shouldn't this function be in the header as well?

Reply:
Both, `ipts_add_companion` and `ipts_remove_companion` are exported in the public header file (`include/linux/ipts-companion.h`) for other modules to use. The other functions are internal use by IPTS itself, so they are exported in the local header file inside `driver/`